### PR TITLE
enable server-printing in tests

### PR DESCRIPTION
**Pull request by juanvallejo**
**Release note**:
```release-note
NONE
```

Depends on https://github.com/kubernetes/kubernetes/pull/55637

Separate pull to test the `--experimental-server-print` flag in test-cmd-util tests introduced in https://github.com/kubernetes/kubernetes/pull/55637

cc @soltysh

### CODE REVIEW

File: hack/lib/test.sh [hunk 1]
```diff
@@ -76,7 +76,7 @@ kube::test::object_assert() {
   local args=${5:-}
 
   for j in $(seq 1 ${tries}); do
-    res=$(eval kubectl get -a "${kube_flags[@]}" ${args} $object -o go-template=\"$request\")
+    res=$(eval kubectl get -a "${kube_flags[@]}" ${args} $object --experimental-server-print=true -o go-template=\"$request\")
```
Comment (deads2k) [inline hack/lib/test.sh]: doing this in combination with a `-o go-template` doesn't do anything

File: hack/make-rules/test-cmd-util.sh [hunk 2]
```diff
@@ -360,7 +360,7 @@ run_pod_tests() {
   kube::test::get_object_assert 'pods/valid-pod' "{{.metadata.namespace}} {{.metadata.name}}" '<no value> valid-pod' "--export=true"
 
   ### Dump current valid-pod POD
-  output_pod=$(kubectl get pod valid-pod -o yaml --output-version=v1 "${kube_flags[@]}")
+  output_pod=$(kubectl get --experimental-server-print=true pod valid-pod -o yaml --output-version=v1 "${kube_flags[@]}")
```
Comment (deads2k) [inline hack/make-rules/test-cmd-util.sh]: same with -o yaml

**Comment by deads2k** [comment]
This doesn't look right.

Why don't you try printing `kubectl get --experimental-server-print --all-namespaces` against all types in discovery?  A separate test.

**Comment by juanvallejo** [comment]
@deads2k thanks for the feedback. Updated pull to introduce separate tests, ensure the output of `get` using experimental print flag matches what is expected

**Review by soltysh — APPROVED** [review]
/lgtm

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by juanvallejo** [comment]
@smarterclayton @soltysh 
The server is returning different columns when requesting a Table for Deployments:

```
$ kubectl get deployments
NAME                         DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
nginx-deployment-resources   3         0         0            0           2h

$ kubectl get deployments --experimental-server-print
NAME                         CREATED AT
nginx-deployment-resources   2018-02-23T18:18:26Z
```

<details>
  <summary>`--v8` of `get deployments --experimental-server-print`</summary>
<code><pre>
...
I0223 16:08:30.612884    2391 round_trippers.go:383] GET http://localhost:8080/apis/mygroup.example.com/v1alpha1
I0223 16:08:30.612958    2391 round_trippers.go:390] Request Headers:
I0223 16:08:30.612971    2391 round_trippers.go:393]     Accept: application/json, */*
I0223 16:08:30.612992    2391 round_trippers.go:393]     User-Agent: kubectl/v1.11.0 (linux/amd64) kubernetes/a2c0fe2
I0223 16:08:30.613489    2391 round_trippers.go:408] Response Status: 200 OK in 0 milliseconds
I0223 16:08:30.613500    2391 round_trippers.go:411] Response Headers:
I0223 16:08:30.613506    2391 round_trippers.go:414]     Content-Type: application/json
I0223 16:08:30.613512    2391 round_trippers.go:414]     Date: Fri, 23 Feb 2018 21:08:30 GMT
I0223 16:08:30.613517    2391 round_trippers.go:414]     Content-Length: 269
I0223 16:08:30.616024    2391 request.go:874] Response Body: {"kind":"APIResourceList","apiVersion":"v1","groupVersion":"mygroup.example.com/v1alpha1","resources":[{"name":"resources","singularName":"resource","namespaced":true,"kind":"Kind","verbs":["delete","deletecollection","get","list","patch","create","update","watch"]}]}
I0223 16:08:30.621164    2391 round_trippers.go:383] GET http://localhost:8080/apis/extensions/v1beta1/namespaces/default/deployments?limit=500
I0223 16:08:30.621179    2391 round_trippers.go:390] Request Headers:
I0223 16:08:30.621187    2391 round_trippers.go:393]     User-Agent: kubectl/v1.11.0 (linux/amd64) kubernetes/a2c0fe2
I0223 16:08:30.621194    2391 round_trippers.go:393]     Accept: application/json;as=Table;v=v1beta1;g=meta.k8s.io, application/json
I0223 16:08:30.622282    2391 round_trippers.go:408] Response Status: 200 OK in 1 milliseconds
I0223 16:08:30.622296    2391 round_trippers.go:411] Response Headers:
I0223 16:08:30.622304    2391 round_trippers.go:414]     Content-Length: 1574
I0223 16:08:30.622317    2391 round_trippers.go:414]     Content-Type: application/json
I0223 16:08:30.622324    2391 round_trippers.go:414]     Date: Fri, 23 Feb 2018 21:08:30 GMT
I0223 16:08:30.622375    2391 request.go:874] Response Body: {"kind":"Table","apiVersion":"meta.k8s.io/v1beta1","metadata":{"selfLink":"/apis/extensions/v1beta1/namespaces/default/deployments","resourceVersion":"227"},"columnDefinitions":[{"name":"Name","type":"string","format":"name","description":"Name must be unique within a namespace. Is required when creating resources, although some resources may allow a client to request the generation of an appropriate name automatically. Name is primarily intended for creation idempotence and configuration definition. Cannot be updated. More info: http://kubernetes.io/docs/user-guide/identifiers#names","priority":0},{"name":"Created At","type":"date","format":"","description":"CreationTimestamp is a timestamp representing the server time when this object was created. It is not guaranteed to be set in happens-before order across separate operations. Clients may not set this value. It is represented in RFC3339 form and is in UTC.\n\nPopulated by the system. Read-only. Null for lists. More info: https://git.k8s.io/community/contr [truncated 550 chars]
I0223 16:08:30.622920    2391 round_trippers.go:383] GET http://localhost:8080/swagger-2.0.0.pb-v1
I0223 16:08:30.622933    2391 round_trippers.go:390] Request Headers:
I0223 16:08:30.622942    2391 round_trippers.go:393]     Accept: application/json, */*
I0223 16:08:30.622949    2391 round_trippers.go:393]     User-Agent: kubectl/v1.11.0 (linux/amd64) kubernetes/a2c0fe2
I0223 16:08:30.628634    2391 round_trippers.go:408] Response Status: 200 OK in 5 milliseconds
I0223 16:08:30.628656    2391 round_trippers.go:411] Response Headers:
I0223 16:08:30.628667    2391 round_trippers.go:414]     Date: Fri, 23 Feb 2018 21:08:30 GMT
I0223 16:08:30.628680    2391 round_trippers.go:414]     Etag: "CEA321EFB38773D6DC2877CB2142D65472892A5EBDB9F5B508BB325975DCA45191951A1614F6D79E96F7B28156353D052D3765B1D7CFEA68DFD9120D150A0D00"
I0223 16:08:30.628692    2391 round_trippers.go:414]     Last-Modified: Thu, 22 Feb 2018 16:04:23 GMT
I0223 16:08:30.628701    2391 round_trippers.go:414]     X-From-Cache: 1
I0223 16:08:30.628709    2391 round_trippers.go:414]     Accept-Ranges: bytes
I0223 16:08:30.628717    2391 round_trippers.go:414]     Content-Type: application/octet-stream
I0223 16:08:30.628725    2391 round_trippers.go:414]     Vary: Accept-Encoding
I0223 16:08:30.723783    2391 request.go:872] Response Body:
...
</pre></code>
</details>

<details>
<summary>`--v8` of `get deployments` with no experimental output</summary>
<code><pre>
I0223 16:12:38.762865    2922 round_trippers.go:383] GET http://localhost:8080/apis/mygroup.example.com/v1alpha1
I0223 16:12:38.762884    2922 round_trippers.go:390] Request Headers:
I0223 16:12:38.762893    2922 round_trippers.go:393]     Accept: application/json, */*
I0223 16:12:38.762904    2922 round_trippers.go:393]     User-Agent: kubectl/v1.11.0 (linux/amd64) kubernetes/a2c0fe2
I0223 16:12:38.764229    2922 round_trippers.go:408] Response Status: 200 OK in 1 milliseconds
I0223 16:12:38.764296    2922 round_trippers.go:411] Response Headers:
I0223 16:12:38.764353    2922 round_trippers.go:414]     Content-Type: application/json
I0223 16:12:38.764381    2922 round_trippers.go:414]     Date: Fri, 23 Feb 2018 21:12:38 GMT
I0223 16:12:38.764399    2922 round_trippers.go:414]     Content-Length: 269
I0223 16:12:38.767930    2922 request.go:874] Response Body: {"kind":"APIResourceList","apiVersion":"v1","groupVersion":"mygroup.example.com/v1alpha1","resources":[{"name":"resources","singularName":"resource","namespaced":true,"kind":"Kind","verbs":["delete","deletecollection","get","list","patch","create","update","watch"]}]}
I0223 16:12:38.773130    2922 round_trippers.go:383] GET http://localhost:8080/apis/extensions/v1beta1/namespaces/default/deployments?limit=500
I0223 16:12:38.773144    2922 round_trippers.go:390] Request Headers:
I0223 16:12:38.773159    2922 round_trippers.go:393]     Accept: application/json
I0223 16:12:38.773165    2922 round_trippers.go:393]     User-Agent: kubectl/v1.11.0 (linux/amd64) kubernetes/a2c0fe2
I0223 16:12:38.774214    2922 round_trippers.go:408] Response Status: 200 OK in 1 milliseconds
I0223 16:12:38.774229    2922 round_trippers.go:411] Response Headers:
I0223 16:12:38.774234    2922 round_trippers.go:414]     Date: Fri, 23 Feb 2018 21:12:38 GMT
I0223 16:12:38.774238    2922 round_trippers.go:414]     Content-Length: 1380
I0223 16:12:38.774242    2922 round_trippers.go:414]     Content-Type: application/json
I0223 16:12:38.774288    2922 request.go:874] Response Body: {"kind":"DeploymentList","apiVersion":"extensions/v1beta1","metadata":{"selfLink":"/apis/extensions/v1beta1/namespaces/default/deployments","resourceVersion":"228"},"items":[{"metadata":{"name":"nginx-deployment-resources","namespace":"default","selfLink":"/apis/extensions/v1beta1/namespaces/default/deployments/nginx-deployment-resources","uid":"f118205a-18c5-11e8-bc4d-507b9dac96e1","resourceVersion":"186","generation":1,"creationTimestamp":"2018-02-23T18:18:26Z","labels":{"name":"nginx-deployment-resources"}},"spec":{"replicas":3,"selector":{"matchLabels":{"name":"nginx"}},"template":{"metadata":{"creationTimestamp":null,"labels":{"name":"nginx"}},"spec":{"containers":[{"name":"nginx","image":"k8s.gcr.io/nginx:test-cmd","ports":[{"containerPort":80,"protocol":"TCP"}],"resources":{},"terminationMessagePath":"/dev/termination-log","terminationMessagePolicy":"File","imagePullPolicy":"IfNotPresent"},{"name":"perl","image":"k8s.gcr.io/perl","resources":{},"terminationMessagePath":"/dev/termination-log","terminati [truncated 356 chars]
I0223 16:12:38.774802    2922 round_trippers.go:383] GET http://localhost:8080/swagger-2.0.0.pb-v1
I0223 16:12:38.774811    2922 round_trippers.go:390] Request Headers:
I0223 16:12:38.774816    2922 round_trippers.go:393]     Accept: application/json, */*
I0223 16:12:38.774821    2922 round_trippers.go:393]     User-Agent: kubectl/v1.11.0 (linux/amd64) kubernetes/a2c0fe2
I0223 16:12:38.781725    2922 round_trippers.go:408] Response Status: 200 OK in 6 milliseconds
I0223 16:12:38.781743    2922 round_trippers.go:411] Response Headers:
I0223 16:12:38.781757    2922 round_trippers.go:414]     Accept-Ranges: bytes
I0223 16:12:38.781764    2922 round_trippers.go:414]     Etag: "CEA321EFB38773D6DC2877CB2142D65472892A5EBDB9F5B508BB325975DCA45191951A1614F6D79E96F7B28156353D052D3765B1D7CFEA68DFD9120D150A0D00"
I0223 16:12:38.781773    2922 round_trippers.go:414]     Last-Modified: Thu, 22 Feb 2018 16:04:23 GMT
I0223 16:12:38.781780    2922 round_trippers.go:414]     Vary: Accept-Encoding
I0223 16:12:38.781788    2922 round_trippers.go:414]     Content-Type: application/octet-stream
I0223 16:12:38.781795    2922 round_trippers.go:414]     Date: Fri, 23 Feb 2018 21:12:38 GMT
I0223 16:12:38.781802    2922 round_trippers.go:414]     X-From-Cache: 1
I0223 16:12:38.895044    2922 request.go:872] Response Body:
</pre></code>
</details>

<br />
I can go ahead and use another resource for the test, but wanted to point this out first.

**Comment by soltysh** [comment]
@juanvallejo yes, use different resource, we can fix that deployments on Monday or over the weekend (if time allows ;))

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by soltysh** [comment]
@juanvallejo it looks like the test-cmd failures are related to your pr, mind addressing them?

**Comment by juanvallejo** [comment]
@soltysh sure, will take a look

**Comment by juanvallejo** [comment]
@soltysh @smarterclayton 
The server is also returning different columns when requesting a Table for HPAs:

```
$ kubectl get hpa
NAME       REFERENCE             TARGETS         MINPODS   MAXPODS   REPLICAS   AGE
frontend   ReplicaSet/frontend   <unknown>/80%   1         1         0          6s

$ kubectl get hpa --experimental-server-print
NAME       CREATED AT
frontend   2018-02-24T23:03:52Z
```

<details>
  <summary>`--v8` of `get deployments --experimental-server-print`</summary>
<code><pre>...

**Comment by smarterclayton** [comment]
Looks like it's missing its table printer.

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by soltysh** [comment]
> Looks like it's missing its table printer.

It looks like there are more types missing those, in which case we're just falling back to default which returns name and creation timestamp. I'm working on adding it to all the missing types.

**Pull request merged**